vivisect/analysis/amd64/emulation.py

```diff
@@ -33,96 +31,62 @@ def prehook(self, emu, op, starteip):
-    3: ('r9',  e_amd64.REG_R9),
-}
-
-arch_bindings = {
-    'msx64call': msx64argnames,
-    'sysvamd64call': sysvamd64argnames,
-    None: [],
-}
-
-def sysvamd64name(idx):
-    ret = sysvamd64argnames.get(idx)
-    if ret == None:
-        name = 'arg%d' % idx
-    else:
-        name, idx = ret
-    return name
-
-def msx64name(idx):
-    ret = msx64argnames.get(idx)
-    if ret == None:
-        name = 'arg%d' % idx
-    else:
-        name, idx = ret
-    return name
-#####
-
 def buildFunctionApi(vw, fva, emu, emumon):
-    
+    '''
+    Builds the function API:
+    * argc - number of arguments
+    * cc   - determines calling convention
+    '''
     argc = 0
     funcargs = []
     callconv = vw.getMeta('DefaultCall')
-    argnames = arch_bindings.get(callconv)
+    cc = emu.getCallingConvention(callconv)
+    argnames = cc.getCallRegArgInfo(emu)
     undefregs = set(emu.getUninitRegUse())
 
+    ### determine argument count, register and stack
+    # determine number of register args
     for argnum in range(len(argnames), 0, -1):
-        argname, argid = argnames[argnum-1]
+        #argname, argid = argnames[argnum-1]
+        argtype, argname, argid = argnames[argnum-1]
         if argid in undefregs:
             argc = argnum
             break
 
+    # calculate the stack argument space used
+    if emumon.stackmax > cc.pad:
+        # stack delta - padding (shadow space, etc)
+        stackargspace = emumon.stackmax - cc.pad
+        targc = cc.getNumRegArgs(emu, MAX_ARGS) + (stackargspace / 8)
+
+        if targc > MAX_ARGS:
+            emumon.logAnomaly(emu, fva, 'Crazy Stack Offset Touched: 0x%.8x' % emumon.stackmax)
+        else:
+            argc = targc
+
+    # add in shadow space for msx64call
     if callconv == 'msx64call':
         # For msx64call there's the shadow space..
-        if emumon.stackmax >= 40:
-            #argc += ((emumon.stackmax - 40) / 8)
-            targc = (emumon.stackmax / 8) - 1
-            if targc > 40:
-                emumon.logAnomaly(emu, fva, 'Crazy Stack Offset Touched: 0x%.8x' % emumon.stackmax)
-                #argc = 0
-            else:
-                argc = targc
-
         # Add the shadow space "locals"
         vw.setFunctionLocal(fva, 8,  LSYM_NAME, ('void *','shadow0'))
         vw.setFunctionLocal(fva, 16, LSYM_NAME, ('void *','shadow1'))
         vw.setFunctionLocal(fva, 24, LSYM_NAME, ('void *','shadow2'))
         vw.setFunctionLocal(fva, 32, LSYM_NAME, ('void *','shadow3'))
 
-        funcargs = [ ('int',msx64name(i)) for i in xrange(argc) ]
-
-    elif callconv == 'sysvamd64call':
-        if emumon.stackmax > 0:
-            targc = (emumon.stackmax / 8) + 6
-            if targc > 40:
-                emumon.logAnomaly(emu, fva, 'Crazy Stack Offset Touched: 0x%.8x' % emumon.stackmax)
-                #argc = 0
-            else:
-                argc = targc
-
-        funcargs = [ ('int',sysvamd64name(i)) for i in xrange(argc) ]
```

Comment:
with current abstraction, only thing left is for msx64call and it's shadow regs.  if we want to make CC classes have an arbitrary update function, the vw.setFunctionLocal() calls could even be placed there to make this fully CC-agnostic.  the current changes use the basic CC def variables to generate the data.  
because i386 has more analysis code, i didn't feel that fully abstracting this out was necessary.